test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,423 @@
+
+        {"f32_ceil", {34, Type::f32, Type::f32}},
+        {"f32_floor", {35, Type::f32, Type::f32}},
+        {"f32_trunc", {36, Type::f32, Type::f32}},
+        {"f32_nearest", {37, Type::f32, Type::f32}},
+        {"f64_ceil", {38, Type::f64, Type::f64}},
+        {"f64_floor", {39, Type::f64, Type::f64}},
+        {"f64_trunc", {40, Type::f64, Type::f64}},
+        {"f64_nearest", {41, Type::f64, Type::f64}},
+    };
+
+    if (const auto it = map.find(name); it != map.end())
+        return it->second;
+    throw std::invalid_argument{"unknown <function>: " + std::string{name}};
+}
+
+constexpr fizzy::Value make_value(Type type, uint64_t bits) noexcept
+{
+    switch (type)
+    {
+    case Type::i32:
+        return fizzy::Value{static_cast<uint32_t>(bits)};
+    case Type::i64:
+        return fizzy::Value{bits};
+    case Type::f32:
+        return fizzy::Value{FP{static_cast<uint32_t>(bits)}.value};
+    case Type::f64:
+        return fizzy::Value{FP{bits}.value};
+    }
+    __builtin_unreachable();
+}
+
+class TypedValue
+{
+    Type type_;
+    fizzy::Value value_;
+
+public:
+    constexpr TypedValue(Type type, fizzy::Value value) noexcept : type_{type}, value_{value} {}
+
+    constexpr TypedValue(Type type, uint64_t bits) noexcept
+      : type_{type}, value_{make_value(type, bits)}
+    {}
+
+    bool eq(uint64_t expected_bits, bool ignore_nans) const
+    {
+        switch (type_)
+        {
+        case Type::i32:
+        case Type::i64:
+            return value_.i64 == expected_bits;
+        case Type::f32:
+        {
+            const FP value{value_.f32};
+            const FP expected{static_cast<uint32_t>(expected_bits)};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                return false;  // Signaling NaN should never happen.
+            }
+            return value == expected;
+        }
+
+        case Type::f64:
+        {
+            const FP value{value_.f64};
+            const FP expected{expected_bits};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                return false;  // Signaling NaN should never happen.
+            }
+            return value == expected;
+        }
+        }
+        __builtin_unreachable();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const TypedValue& x)
+    {
+        os << std::hex << std::uppercase << std::setfill('0');
+        switch (x.type_)
+        {
+        case Type::i32:
+            return os << std::setw(8) << x.value_.as<uint32_t>();
```

Comment:
I'm not a fan on `std::hex` and `std::setw`, there's always some mess with it in Solidity at least. Could we use a new version of the `hex` helper we have in `test/utils/hex`?

Reply:
The `setw` is fine as it only applies to the next number.
The `hex` is also fine because we display all in hex anyway.

Reply:
I know it is "fine" but somehow there are differences with these on different architectures. Solidity has weird issues with these time to time, between the different platforms.

Reply:
I'm restoring the ostream state at exit, but still `std::hex` and others are used (no arguments to assume they don't work).